Track emitted events on model

AlanCoding opened the pull request:
This saves the number of events generated by a job so we can make intelligent decisions based on that later.

WIP because failed inventory updates are (very unexpectedly) not following the code path in the `finally:`

AlanCoding committed:
Track emitted events on model

AlanCoding commented:
Alright, I've tested all the models and they're working now, in that events get processed, and if I force events to not all be processed, the field reflects this.

chrismeyersfsu commented on awx/api/serializers.py:
```diff
@@ -643,13 +643,14 @@ def to_representation(self, obj):
 
 class UnifiedJobSerializer(BaseSerializer):
     show_capabilities = ['start', 'delete']
+    events_processed = serializers.SerializerMethodField()
```
Can you make this Boolean so options is accurate?

AlanCoding replied:
I was just googling for this... I don't know how. If I make a boolean field, then DRF expectation is that it's present on the model.

I guess that forces our hand, and we should just cave and add it as a model property?

chrismeyersfsu commented on awx/api/views.py:
```diff
@@ -145,6 +145,15 @@ def destroy(self, request, *args, **kwargs):
             raise PermissionDenied(detail=_("Cannot delete running job resource."))
+        elif obj.emitted_events > obj.get_event_queryset().count():
+            # Prohibit deletion if job events are still coming in
+            if obj.modified + dateutil.relativedelta.relativedelta(minutes=1) < now():
```
consider moving this logic to UnifiedJob or as a mixin on UnifiedJob

chrismeyersfsu commented on awx/api/serializers.py:
```diff
@@ -667,6 +668,11 @@ def get_types(self):
         else:
             return super(UnifiedJobSerializer, self).get_types()
 
+    def get_events_processed(self, obj):
```
Incorporate the timeout that's in the view here ? Return `True` if timeout reached.

chrismeyersfsu reviewed (commented):
I like the way you passed the event count :+1:

AlanCoding commented:
Okay, field now just in the GET options as

```json
            "events_processed": {
                "type": "boolean",
                "label": "Events processed",
                "help_text": "Indicates whether all of the events generated by this unified job have been saved to the database."
            },
```

AlanCoding commented on awx/main/models/events.py:
```diff
@@ -570,6 +570,12 @@ def create_from_data(self, **kwargs):
+        Needed for __unicode__
+        '''
+        return self.event
+
```
@ryanpetrello I was getting this just spelunking around in the console

```
In [3]: iu.get_event_queryset()
Out[3]: ---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/venv/awx/lib/python2.7/site-packages/IPython/core/formatters.pyc in __call__(self, obj)
    670                 type_pprinters=self.type_printers,
    671                 deferred_pprinters=self.deferred_printers)
--> 672             printer.pretty(obj)
    673             printer.flush()
    674             return stream.getvalue()

/venv/awx/lib/python2.7/site-packages/IPython/lib/pretty.pyc in pretty(self, obj)
    381                             if callable(meth):
    382                                 return meth(obj, self, cycle)
--> 383             return _default_pprint(obj, self, cycle)
    384         finally:
    385             self.end_group()

/venv/awx/lib/python2.7/site-packages/IPython/lib/pretty.pyc in _default_pprint(obj, p, cycle)
    501     if _safe_getattr(klass, '__repr__', None) not in _baseclass_reprs:
    502         # A user-provided repr. Find newlines and replace them with p.break_()
--> 503         _repr_pprint(obj, p, cycle)
    504         return
    505     p.begin_group(1, '<')

/venv/awx/lib/python2.7/site-packages/IPython/lib/pretty.pyc in _repr_pprint(obj, p, cycle)
    699     """A pprint that just redirects to the normal repr function."""
    700     # Find newlines and replace them with p.break_()
--> 701     output = repr(obj)
    702     for idx,output_line in enumerate(output.splitlines()):
    703         if idx:

/venv/awx/lib/python2.7/site-packages/django/db/models/query.pyc in __repr__(self)
    227         if len(data) > REPR_OUTPUT_SIZE:
    228             data[-1] = "...(remaining elements truncated)..."
--> 229         return '<%s %r>' % (self.__class__.__name__, data)
    230 
    231     def __len__(self):

/venv/awx/lib/python2.7/site-packages/django/db/models/base.pyc in __repr__(self)
    588     def __repr__(self):
    589         try:
--> 590             u = six.text_type(self)
    591         except (UnicodeEncodeError, UnicodeDecodeError):
    592             u = '[Bad Unicode data]'

/awx_devel/awx/main/models/events.pyc in __unicode__(self)
    547 
    548     def __unicode__(self):
--> 549         return u'%s @ %s' % (self.get_event_display(), self.created.isoformat())
    550 
    551     @classmethod

AttributeError: 'InventoryUpdateEvent' object has no attribute 'get_event_display'
```

ryanpetrello replied:
Interesting.  This must be some sort of fallout from my job event model refactoring a few weeks back.  Good catch.

AlanCoding commented on awx/api/views.py:
```diff
@@ -145,6 +145,15 @@ def destroy(self, request, *args, **kwargs):
             raise PermissionDenied(detail=_("Cannot delete running job resource."))
+        elif not obj.events_processed:
+            # Prohibit deletion if job events are still coming in
+            if obj.modified + dateutil.relativedelta.relativedelta(minutes=1) < now():
```
actually, maybe it would be better to use `obj.finished` here instead.

And @chrismeyersfsu I'm not really a fan of putting the timedelta logic inside of the model's method. It would make the field confusing to use. If the job has been finished for a long while and the events aren't fully processed, or if this is happening on the regular, that would be a bad state of the server, and it's better to show that as opposed to hiding it.

ryanpetrello commented on awx/api/views.py:
```diff
@@ -145,6 +145,15 @@ def destroy(self, request, *args, **kwargs):
             raise PermissionDenied(detail=_("Cannot delete running job resource."))
+        elif not obj.events_processed:
+            # Prohibit deletion if job events are still coming in
+            if obj.finished + dateutil.relativedelta.relativedelta(minutes=1) < now():
```
@AlanCoding my brain had some serious trouble parsing this, and I'm not sure why.  Shouldn't this be _greater than now_ rather than _less than now_?

if `<finished_stamp> + <one minute>` *is after* `<right now>`, then tell them the job is still processing?

This might be easier to read if it were presented in the opposite direction:

```python
if now() < obj.finished + dateutil.relativedelta.relativedelta(minutes=1):
   # a minute has not passed
```

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -911,6 +915,29 @@ def result_stdout_text(self, value):
+        return self.event_class.objects.filter(**{self.event_parent_key: self.id})
+
+    @property
+    def events_processed(self):
```
Could we rename this property/field to make it clear it's a boolean?  My brain read this property name expecting back a count/integer.  Maybe something like `event_processing_finished` ?

AlanCoding replied:
The name here has to be the same as what we're returning in the API.

That said, yes, I'm in favor of a better name for the field. That's a little long though...

 - events_have_processed
 - finished_processing
 - finalized
 - can_haz_events

ryanpetrello replied:
`actually_finished` 😂

ryanpetrello replied:
I like `events_have_processed`

AlanCoding replied:
I might stereotype `events_have_processed` as an integer. I think `event_processing_finished` is a little more bool-ish, and is my current favorite.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -966,29 +993,21 @@ def result_stdout_raw_handle(self, enforce_max_bytes=True):
                     "copy (select stdout from {} where {}={} order by start_line) to stdout".format(
-                        tablename + 'event',
-                        related_name,
+                        self._meta.db_table + 'event',
```
👏

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1001,14 +1002,11 @@ def run(self, pk, isolated_host=None, **kwargs):
-                                instance.log_format, stdout_handle._event_ct)
-                else:
-                    logger.info('%s finished running', instance.log_format)
+                event_ct = getattr(stdout_handle, '_event_ct', 0)
```
What actually sets this?  Is this diff missing some changes?

ryanpetrello replied:
Oh, the `_event_ct` predates this on the `OutputEventFilter`: https://github.com/ansible/awx/blob/devel/awx/main/utils/common.py#L979.  Nice 😄

AlanCoding replied:
Yes, this is a change we saw coming.

Personally, I have already found substantial value in the log message. There were all kinds of cluster debugging issue where I would search for this log, and then go in the API to `/jobs/N/job_events/` to compare to the number of events. When the instances are coming up/down unpredictably it's not surprising to _lose_ events. To the extent of me using that log, I've found it to be pretty reliable.

This change is automating that task and putting it in the API.

ryanpetrello reviewed (approved):
Minus a few nits in my comments, I really like this change overall - I think this will be really helpful (even beyond the purpose of blocking deletions during processing).

ryanpetrello commented:
@AlanCoding another thought I had: it might be nice to add a test somewhere in here:

https://github.com/ansible/awx/blob/devel/awx/main/tests/unit/test_tasks.py

...ala...

https://github.com/ansible/awx/blob/devel/awx/main/tests/unit/test_tasks.py#L306

...which makes an assertion about `update_model(emitted_events=N)` being called with the proper line/event count.

ryanpetrello commented on awx/api/views.py:
```diff
@@ -145,6 +145,16 @@ def destroy(self, request, *args, **kwargs):
+                return Response({"error": _("Job has not finished processing events.")},
+                                status=status.HTTP_400_BAD_REQUEST)
+            else:
+                # if it has been > 1 minute, events are probably lost
```
Should we _only_ warn in this way if `obj.finished`?  What if you deleted a job that wasn't finished yet, but was in some other state (like `pending`)?

AlanCoding replied:
Yes. In `tasks.py`, we have a line

```python
instance = self.update_model(pk, status=status, result_traceback=tb,
```

This is the canonical finalization save for a job. This should be the first and only time that a job's status changes from an active state to a finalized state.

This new field was piggybacked onto this save.

In the `.save` method for unified jobs, it also tacks on `finished=now()` if the status changed to a finalized state. That means that the finished time gets saved at the same time that the event count does.

I'm worried about preventing false-positives here, and not false-negatives. If the job is still running, then its deletion should already be blocked by other logic. If that logic fails, then I don't want _this_ code to throw a traceback and confuse the issue.

ryanpetrello replied:
👍

AlanCoding committed:
Namechange events_processed -> event_processing_finished

from PR review, also adding tests to assert that the
value is passed from the stdout_handle to the UnifiedJob
object on finalization of job run in tasks.py

ryanpetrello commented:
@AlanCoding :shipit:
